### ISSUE TITLE
`upstream` inspection has started erroneously failing for some packages.

### ISSUE THREAD

**Issue by ckelleyRH** [rpminspect/rpminspect#1461]
**Describe the bug**
We are seeing multiple packages that are failing the `upstream` inspection. In one case, all that changed was a single patch, and multiple source files were flagged as new.

**To Reproduce**
Working: https://artifacts.osci.redhat.com/testing-farm/42cd704d-24cf-4e4c-9321-0db54acc5a38/
Failing: https://artifacts.osci.redhat.com/testing-farm/20a15057-7cbb-4f5e-9c4d-379207e757b7/

Further examples of failure in the ticket: https://issues.redhat.com/browse/OSCI-7809

**Expected behavior**
`upstream` inspection passes in cases where no source updates have occurred, as before.
**Environment:**
RHEL 8 and 9 x86_64.

**Comment by dcantrell** [comment]
I am unable to download task 65763544 to reproduce the problem locally.  Is there another build that demonstrates the failure in the upstream inspection?

**Comment by ckelleyRH** [comment]
66803893 allegedly also affected.

**Comment by dcantrell** [comment]
So I don't know what's happening here.  I cannot reproduce it locally at all.  The upstream inspection keeps passing.

The upstream inspection only works on SRPM files.  It compares the list of Source files from the before build with the after build.  If any of those are different and the package is a rebased build and is on the rebaseable list, then the results are reported at the INFO level.  If it is a rebase but the package is not on the rebaseable list, the results are reported at the VERIFY level.

What is considered a rebase?  The package names match but the version numbers do not.  A general update to a package is considered when the package and version match but the release number does not.

In the passing example you give, the before build is python3-3.6.8-62.el8_10 and the after build is python3-3.6.8-67.el8_10.  Both of these have RPMTAG_NAME as python3 and RPMTAG_VERSION as 3.6.8, so it is not considered a rebase build.  Any changes found by the upstream inspection are reported as INFO.

The failing example has python3-3.6.8-67.el8_10 as the before build and task 65763544 as the after build.  But I couldn't find that task, so I was unable to see what packages were there.

And task 66803893 is a vim build, so I can't compare that with python3.

Can you give me a before and after build where this problem is showing up and that I can access so I can work on it locally?

**Comment by msrb** [comment]
@dcantrell Those tasks are scratch builds -- they still exists in Brew, but it's possible that Brew is doing some soft-cleaning before the tasks are eventually garbage collected completely.

Here's the latest one. This one still exists and can be downloaded:

```
$ rpminspect-redhat --verbose --arches=x86_64,noarch,src -T upstream --release=el8 --profile=rhel-8 pcs-0.10.18-2.el8_10.3 66885739
=> pcs-0.10.18-2.el8_10.4.src.rpm                            [#############################################################]
=> pcs-0.10.18-2.el8_10.4.src.rpm                            [#############################################################]
=> pcs-0.10.18-2.el8_10.4.x86_64.rpm                         [#############################################################]
=> pcs-snmp-0.10.18-2.el8_10.4.x86_64.rpm                    [#############################################################]
Downloading RPM build pcs-0.10.18-2.el8_10.3
=> pcs-0.10.18-2.el8_10.3.src.rpm                            [#############################################################]
=> pcs-0.10.18-2.el8_10.3.x86_64.rpm                         [#############################################################]
=> pcs-snmp-0.10.18-2.el8_10.3.x86_64.rpm                    [#############################################################]
rpminspect: *** archive_read_extract: Hard-link target '/var/tmp/rpminspect/scratch-66885739.5ZzMqC/root/before/x86_64/(null)' does not exist.
rpminspect: *** archive_read_extract: Hard-link target '/var/tmp/rpminspect/scratch-66885739.5ZzMqC/root/after/x86_64/(null)' does not exist.
Skipping abidiff inspection...       skip
Skipping addedfiles inspection...    skip
Skipping annocheck inspection...     skip
Skipping arch inspection...          skip
Skipping badfuncs inspection...      skip
Skipping capabilities inspection...  skip
Skipping changedfiles inspection...  skip
Skipping changelog inspection...     skip
Skipping config inspection...        skip
Skipping debuginfo inspection...     skip
Skipping desktop inspection...       skip
Skipping disttag inspection...       skip
Skipping doc inspection...           skip
Skipping dsodeps inspection...       skip
Skipping elf inspection...           skip
Skipping emptyrpm inspection...      skip
Skipping files inspection...         skip
Skipping filesize inspection...      skip
Skipping javabytecode inspection...  skip
Skipping kmidiff inspection...       skip
Skipping kmod inspection...          skip
Skipping license inspection...       skip
Skipping lostpayload inspection...   skip
Skipping lto inspection...           skip
Skipping manpage inspection...       skip
Skipping metadata inspection...      skip
Skipping modularity inspection...    skip
Skipping movedfiles inspection...    skip
Skipping ownership inspection...     skip
Skipping patches inspection...       skip
Skipping pathmigration inspection... skip
Skipping permissions inspection...   skip
Skipping politics inspection...      skip
Skipping removedfiles inspection...  skip
Skipping rpmdeps inspection...       skip
Skipping runpath inspection...       skip
Skipping shellsyntax inspection...   skip
Skipping specname inspection...      skip
Skipping subpackages inspection...   skip
Skipping symlinks inspection...      skip
Skipping types inspection...         skip
Skipping udevrules inspection...     skip
Skipping unicode inspection...       skip
Running upstream inspection...       FAIL
Skipping virus inspection...         skip
Skipping xml inspection...           skip

diagnostics:
------------
1) Version information for libraries and programs used by rpminspect as well as storage requirements.  This result is for
informational and diagnostic purposes only. 

Result: DIAGNOSTICS
Details:
rpminspect version 2.1
zlib version 1.3.1.zlib-ng
libmagic version 546
libclamav version 1.4.2
librpm version 4.20.0
libxml version 21209
json-c version 0.18
libcurl version 8.11.1 (OpenSSL/3.2.4 zlib/1.3.1.zlib-ng brotli/1.1.0 libidn2/2.3.7 libpsl/0.21.5 libssh/0.11.1/openssl/zlib nghttp2/1.64.0 OpenLDAP/2.6.9)
libarchive version 3.7.7 (zlib/1.3.1.zlib-ng liblzma/5.6.3 bz2lib/1.0.8 liblz4/1.10.0 libzstd/1.5.6)
libyaml version 0.2.5
OpenSSL version 3.2.4 11 Feb 2025
xmlrpc-c version 1.60.4
msgunfmt (GNU gettext-tools) 0.23.1
annocheck version 12.88.
abidiff version 2.6.0
kmidiff version 2.6.0
udevadm version 257

Space required to download artifacts: 25039084 bytes (24 MiB)
Space required to unpack artifacts: 49543420 bytes (47 MiB)



2) Command line arguments used to invoke rpminspect. 

Result: DIAGNOSTICS
Details:
/usr/bin/rpminspect -c /usr/share/rpminspect/redhat.yaml --verbose --arches=x86_64,noarch,src -T upstream --release=el8 --profile=rhel-8 pcs-0.10.18-2.el8_10.3 66885739


3) Before Build 

Result: DIAGNOSTICS
Details:
pcs-0.10.18-2.el8_10.3


4) After Build 

Result: DIAGNOSTICS
Details:
66885739

```

It appears that the after build SRPM is being download twice (?) And then the following lines look suspicious:

```
rpminspect: *** archive_read_extract: Hard-link target '/var/tmp/rpminspect/scratch-66885739.5ZzMqC/root/before/x86_64/(null)' does not exist.
rpminspect: *** archive_read_extract: Hard-link target '/var/tmp/rpminspect/scratch-66885739.5ZzMqC/root/after/x86_64/(null)' does not exist.
```

**Comment by dcantrell** [comment]
So with this information, I am able to recreate the problem.  In the case of the pcs packages, the failure reported by rpminpect was caused by tornado-6.1.0.tar.gz being replaced with tornado-6.1.0.pcs.1.tar.gz, *BUT* the package version number remained the same.

The upstream inspection determines if you are doing a maintenance inspection or a full inspection by way of the hospital.  When two builds have the same name and version number but the release changes, rpminspect considers that a maintenance update.  It expects all of the Sources in the SRPM to be the same, only things added.  If the package version also changes but the name remains the same, rpminspect considers that a rebase build.  Changes between Source files in a rebase comparison, rpminspect will report those at the INFO level because during a rebase, anything is expect.

There are a couple of options here:

- You can waive the inspection result and let it through.  There is now way to alter behavior through the rpminspect.yaml file.
- Modify the newer build so the package version number changes.  This will cause rpminspect to treat the job comparison as a rebase.

**Comment by dcantrell** [comment]
OK, my misunderstanding then.  I tried to reproduce the problem locally and noticed it did the correct thing and reported the changing tornado source archives at the VERIFY level.  But looking at the link above, I do see the other source files mentioned.  I am unable to reproduce that locally with the latest rpminspect

@msrb where is RHEL on Gitlab getting rpminspect from?  Is it running the Copr builds like it should be?  https://copr.fedorainfracloud.org/coprs/dcantrell/rpminspect/  Can I see the code that invokes rpminspect in RHEL on Gitlab?

**Comment by msrb** [comment]
@dcantrell I suspect that this will be the root cause of all these issues:
https://github.com/rpminspect/rpminspect/issues/1464

RHEL on GitLab is getting rpminspect from here: https://copr.fedorainfracloud.org/coprs/g/osci/rpminspect/
It's a SRPM rebuild from your COPR repo (with some delays, i.e. we don't use nightlies).

The exact command line is visible in the html report:
https://artifacts.osci.redhat.com/testing-farm/1a14bc30-8c9a-4ca0-8ab3-1e0cb64270dd/
Under "Diagnostics". In this case, it was:
```
Command line arguments used to invoke rpminspect.
/usr/bin/rpminspect -c /usr/share/rpminspect/redhat.yaml --workdir /var/ARTIFACTS/work-rpminspectd3fro3w1/rpminspect/discover/default-0/tests/workdir/ --format=json --output=/var/ARTIFACTS/work-rpminspectd3fro3w1/rpminspect/execute/data/guest/default-0/rpminspect-1/data/result.json --verbose --arches=x86_64,noarch,src --release=el8 --profile=rhel-8 pcs-0.10.18-2.el8_10.3 66885739
```

**Comment by msrb** [comment]
The latest example:

report: https://artifacts.osci.redhat.com/testing-farm/1af2859e-cf20-4cb7-8127-023668d83ada/
verbose log: https://artifacts.osci.redhat.com/testing-farm/1af2859e-cf20-4cb7-8127-023668d83ada/work-rpminspect1tvtl9qe/rpminspect/execute/data/guest/default-0/rpminspect-1/data/verbose.log

```
$ rpminspect-redhat --verbose -T upstream --release=el8 --profile=rhel-8 dotnet8.0-8.0.114-1.el8_10 66955157
```

**Comment by msrb** [comment]
And here's even smaller reproducer:

```
$ rpminspect-redhat --verbose -T upstream --arches=src --release=el8 --profile=rhel-8 libxml2-2.9.7-18.el8_10.2 66949389
=> libxml2-2.9.7-19.el8_10.src.rpm                    [######################################################]
=> libxml2-2.9.7-19.el8_10.src.rpm                    [######################################################]
Downloading RPM build libxml2-2.9.7-18.el8_10.2
=> libxml2-2.9.7-18.el8_10.2.src.rpm                  [######################################################]
Skipping abidiff inspection...       skip
Skipping addedfiles inspection...    skip
Skipping annocheck inspection...     skip

<snip>

upstream:
---------
1) New upstream source file `libxml2-2.9.7.tar.gz` appeared 

Result: VERIFY
Waiver Authorization: Anyone

Suggested Remedy: Unexpected changed source archive content. The version of the package did not change between
builds, but the source archive content did. This may be deliberate, but needs inspection. If this change is
expected, update the rebaseable exception list for this product release and send a patch to the project that
owns the rpminspect data files.

2) Source file `libxml2-2.9.7.tar.gz` removed 

Result: VERIFY
Waiver Authorization: Anyone

Suggested Remedy: Unexpected changed source archive content. The version of the package did not change between
builds, but the source archive content did. This may be deliberate, but needs inspection. If this change is
expected, update the rebaseable exception list for this product release and send a patch to the project that
owns the rpminspect data files.

```
